src/renderers/dom/client/wrappers/ReactDOMInput.js

```diff
@@ -144,6 +146,10 @@ var ReactDOMInput = {
       listeners: null,
       onChange: _handleChange.bind(inst),
     };
+
+    if (__DEV__) {
+      inst._wrapperState.controlled = props.value !== undefined;
```

Comment:
If `type="checkbox"` then the controlled component would use the `checked` prop instead of the `value` prop.